diff --git a/addr_gonla.go b/addr_gonla.go
new file mode 100644
index 0000000..f76e094
--- /dev/null
+++ b/addr_gonla.go
@@ -0,0 +1,9 @@
+// -*- coding: utf-8 -*-
+
+package netlink
+
+import ()
+
+func AddrDeserialize(m []byte) (addr Addr, family, index int, err error) {
+	return parseAddr(m)
+}
diff --git a/nl/mpls_gonla.go b/nl/mpls_gonla.go
new file mode 100644
index 0000000..bb11deb
--- /dev/null
+++ b/nl/mpls_gonla.go
@@ -0,0 +1,5 @@
+// -*- coding: utf-8 -*-
+
+package nl
+
+const RTNLGRP_MPLS_ROUTE = 27
diff --git a/nl/nl_linux_gonla.go b/nl/nl_linux_gonla.go
new file mode 100644
index 0000000..a193981
--- /dev/null
+++ b/nl/nl_linux_gonla.go
@@ -0,0 +1,21 @@
+package nl
+
+import (
+	"golang.org/x/sys/unix"
+)
+
+func (s *NetlinkSocket) SetReceiveBuffer(bufsize int) error {
+	return unix.SetsockoptInt(int(s.fd), unix.SOL_SOCKET, unix.SO_RCVBUF, bufsize)
+}
+
+func (s *NetlinkSocket) GetReceiveBuffer() (int, error) {
+	return unix.GetsockoptInt(int(s.fd), unix.SOL_SOCKET, unix.SO_RCVBUF)
+}
+
+func (s *NetlinkSocket) SetSendBuffer(bufsize int) error {
+	return unix.SetsockoptInt(int(s.fd), unix.SOL_SOCKET, unix.SO_SNDBUF, bufsize)
+}
+
+func (s *NetlinkSocket) GetSendBuffer() (int, error) {
+	return unix.GetsockoptInt(int(s.fd), unix.SOL_SOCKET, unix.SO_SNDBUF)
+}
diff --git a/nl/syscall.go b/nl/syscall.go
index 0ff1bba..31382cb 100644
--- a/nl/syscall.go
+++ b/nl/syscall.go
@@ -47,6 +47,7 @@ const (
 )
 
 const (
+	RTA_VIA        = 0x12
 	RTA_NEWDST     = 0x13
 	RTA_ENCAP_TYPE = 0x15
 	RTA_ENCAP      = 0x16
diff --git a/route.go b/route.go
index 58ff1af..5a24551 100644
--- a/route.go
+++ b/route.go
@@ -56,7 +56,7 @@ func (r Route) String() string {
 		elems = append(elems, fmt.Sprintf("Ifindex: %d", r.LinkIndex))
 	}
 	if r.MPLSDst != nil {
-		elems = append(elems, fmt.Sprintf("Dst: %d", r.MPLSDst))
+		elems = append(elems, fmt.Sprintf("Dst: %d", *r.MPLSDst))
 	} else {
 		elems = append(elems, fmt.Sprintf("Dst: %s", r.Dst))
 	}
diff --git a/route_gonla.go b/route_gonla.go
new file mode 100644
index 0000000..08d5709
--- /dev/null
+++ b/route_gonla.go
@@ -0,0 +1,10 @@
+// -*- coding: utf-8 -*-
+
+package netlink
+
+import ()
+
+func RouteDeserialize(m []byte) (*Route, error) {
+	route, err := deserializeRoute(m)
+	return &route, err
+}
diff --git a/route_linux.go b/route_linux.go
index 93fb87a..5773ff2 100644
--- a/route_linux.go
+++ b/route_linux.go
@@ -793,6 +793,8 @@ func deserializeRoute(m []byte) (Route, error) {
 		switch attr.Attr.Type {
 		case unix.RTA_GATEWAY:
 			route.Gw = net.IP(attr.Value)
+		case nl.RTA_VIA:
+			route.Gw = net.IP(attr.Value[2:])
 		case unix.RTA_PREFSRC:
 			route.Src = net.IP(attr.Value)
 		case unix.RTA_DST:
@@ -839,6 +841,8 @@ func deserializeRoute(m []byte) (Route, error) {
 					switch attr.Attr.Type {
 					case unix.RTA_GATEWAY:
 						info.Gw = net.IP(attr.Value)
+					case nl.RTA_VIA:
+						info.Gw = net.IP(attr.Value[2:])
 					case nl.RTA_NEWDST:
 						var d Destination
 						switch msg.Family {
